src/random/mod.rs

```diff
@@ -1,12 +1,22 @@
 //! Support for build-in RNGs
 
-#[derive(Debug)]
+#[derive(Copy, Clone, Debug)]
 /// Used to obtain random numbers using x86_64's RDRAND opcode
 pub struct RdRand(());
 
+mod private {
+    /// Trait for types that can be returned by the RDRAND instruction
+    pub trait RdRandPrimitive: Sized {}
+    impl RdRandPrimitive for u16 {}
+    impl RdRandPrimitive for u32 {}
+    #[cfg(target_arch = "x86_64")]
+    impl RdRandPrimitive for u64 {}
+}
+use private::RdRandPrimitive;
```

Comment:
So RdRandPrimitive is okay this way because it is still technically a public thing, just inside a non-public module. I've seen this done in other crates, and also recommended on the rust subreddit as a way to prevent implementing this trait outside of this crate.

Essentially, if RdRandPrimitive is accessible to users of this crate, they could write something like `struct Thing(u64); impl RdRandPrimitive for Thing{}`. This specific example is okay, but if they changed it to `struct Thing(u128); impl RdRandPrimitive for Thing{}`, then that is NOT okay, but still compiles without warning. The latter is not okay because rdrand does not output a 128bit value, so part of `Thing` will be would end up being non-random.

The solutions to this (maybe there's more, just what I could think of) are to either make the trait unsafe (and rely on the implementor to understand how to use it correctly), or put the trait in a private module so the user cannot implement it, and can't do something dumb like the example above.

Reply:
I didn't know that this was a common thing. It still seems hacky and leads to an ugly API documentation:

![image](https://user-images.githubusercontent.com/1131315/52519623-6d12b000-2c5e-11e9-987c-f3cec323561d.png)

There is no way for the user to see for which types `RdRandPrimitive` is implemented because there's no documentation for it as it lives in a private module.

Reply:
I can make documentation more explicit, or pull the trait out of the private module and use a private "Sealed" pattern like they do in in link.